File: src/kernels/activate.rs
use prelude::*;
use kernels::ffi::*;

use densearray::prelude::*;

pub struct ActivateKernel {
  batch_sz: usize,
  dim:      usize,
  act_kind: ActivationKind,
}

impl ActivateKernel {
  pub fn new(batch_sz: usize, dim: usize, act_kind: ActivationKind) -> ActivateKernel {
    ActivateKernel{
      batch_sz: batch_sz,
      dim:      dim,
      act_kind: act_kind,
    }
  }

  pub fn forward(&mut self, batch_sz: usize, in_buf: &[f32], out_buf: &mut [f32]) {
    assert!(batch_sz <= self.batch_sz);
    match self.act_kind {
      ActivationKind::Identity => {
        out_buf.copy_from_slice(in_buf);
      }
      ActivationKind::Rect => {
        unsafe { neuralops_rect_fwd(
            batch_sz,
            self.dim,
            in_buf.as_ptr(),
            out_buf.as_mut_ptr(),
        ) };
      }
      ActivationKind::LeakyRect(_) => {
        unimplemented!();
      }
      ActivationKind::Logistic => {
        unsafe { neuralops_logistic_fwd(
            batch_sz,
            self.dim,
            in_buf.as_ptr(),
            out_buf.as_mut_ptr(),
        ) };
      }
      ActivationKind::Tanh => {
        unimplemented!();
      }
    }
  }

  pub fn backward(&mut self, batch_sz: usize, out_buf: &[f32], out_grad: &[f32], in_grad: &mut [f32]) {
    match self.act_kind {
      ActivationKind::Identity => {
        in_grad.copy_from_slice(out_grad);
      }
      ActivationKind::Rect => {
        unsafe { neuralops_rect_bwd(
            batch_sz,
            self.dim,
            out_buf.as_ptr(),
            out_grad.as_ptr(),
            in_grad.as_mut_ptr(),
        ) };
      }
      ActivationKind::LeakyRect(_) => {
        unimplemented!();
      }
      ActivationKind::Logistic => {
        unsafe { neuralops_logistic_bwd(
            batch_sz,
            self.dim,
            out_buf.as_ptr(),
            out_grad.as_ptr(),
            in_grad.as_mut_ptr(),
        ) };
      }
      ActivationKind::Tanh => {
        unimplemented!();
      }
    }
  }
}

pub struct ParallelActivateKernel {
  batch_sz: usize,
  dim:      usize,
  act_kind: ActivationKind,
}

impl ParallelActivateKernel {
  pub fn new(batch_sz: usize, dim: usize, act_kind: ActivationKind) -> ParallelActivateKernel {
    ParallelActivateKernel{
      batch_sz: batch_sz,
      dim:      dim,
      act_kind: act_kind,
    }
  }

  pub fn forward(&mut self, batch_sz: usize, in_buf: &[f32], out_buf: &mut [f32]) {
    assert!(batch_sz <= self.batch_sz);
    match self.act_kind {
      ActivationKind::Identity => {
        //out_buf.copy_from_slice(in_buf);
        out_buf.reshape_mut(batch_sz * self.dim).parallel_copy(in_buf.reshape(batch_sz * self.dim));
      }
      ActivationKind::Rect => {
        unsafe { neuralops_omp_rect_fwd(
            batch_sz,
            self.dim,
            in_buf.as_ptr(),
            out_buf.as_mut_ptr(),
        ) };
      }
      ActivationKind::LeakyRect(_) => {
        unimplemented!();
      }
      ActivationKind::Logistic => {
        unsafe { neuralops_omp_logistic_fwd(
            batch_sz,
            self.dim,
            in_buf.as_ptr(),
            out_buf.as_mut_ptr(),
        ) };
      }
      ActivationKind::Tanh => {
        unimplemented!();
      }
    }
  }

  pub fn backward(&mut self, batch_sz: usize, out_buf: &[f32], out_grad: &[f32], in_grad: &mut [f32]) {
    match self.act_kind {
      ActivationKind::Identity => {
        //in_grad.copy_from_slice(out_grad);
        in_grad.reshape_mut(batch_sz * self.dim).parallel_copy(out_grad.reshape(batch_sz * self.dim));
      }
      ActivationKind::Rect => {
        unsafe { neuralops_omp_rect_bwd(
            batch_sz,
            self.dim,
            out_buf.as_ptr(),
            out_grad.as_ptr(),
            in_grad.as_mut_ptr(),
        ) };
      }
      ActivationKind::LeakyRect(_) => {
        unimplemented!();
      }
      ActivationKind::Logistic => {
        unsafe { neuralops_omp_logistic_bwd(
            batch_sz,
            self.dim,
            out_buf.as_ptr(),
            out_grad.as_ptr(),
            in_grad.as_mut_ptr(),
        ) };
      }
      ActivationKind::Tanh => {
        unimplemented!();
      }
    }
  }
}

File: extkernels/activate.c
#include <math.h>
#include <stdint.h>
#include <stdlib.h>

void neuralops_rect_fwd(
    size_t batch_sz,
    size_t dim,
    const float *in_buf,
    float *out_buf)
{
  for (size_t p = 0; p < batch_sz * dim; p += 1) {
    float x = in_buf[p];
    if (x > 0.0f) {
      out_buf[p] = x;
    } else {
      out_buf[p] = 0.0f;
    }
  }
}

void neuralops_rect_bwd(
    size_t batch_sz,
    size_t dim,
    const float *in_buf,
    const float *out_grad,
    float *in_grad)
{
  for (size_t p = 0; p < batch_sz * dim; p += 1) {
    float x = in_buf[p];
    if (x > 0.0f) {
      in_grad[p] = out_grad[p];
    } else {
      in_grad[p] = 0.0f;
    }
  }
}

void neuralops_logistic_fwd(
    size_t batch_sz,
    size_t dim,
    const float *in_buf,
    float *out_buf)
{
  for (size_t p = 0; p < batch_sz * dim; p += 1) {
    float x = in_buf[p];
    out_buf[p] = 1.0 / (1.0 + expf(-x));
  }
}

void neuralops_logistic_bwd(
    size_t batch_sz,
    size_t dim,
    const float *in_buf,
    const float *out_grad,
    float *in_grad)
{
  for (size_t p = 0; p < batch_sz * dim; p += 1) {
    float x = in_buf[p];
    float z = expf(-x);
    float y = 1.0 / (1.0 + z);
    in_grad[p] = out_grad[p] * z * y * y;
  }
}
